lib/unparser/comments.rb

```diff
@@ -102,10 +102,8 @@ def take_before(node, source_part)
     # @api private
     #
     def self.source_range(node, part)
-      location = node.location
-      if location && location.respond_to?(part)
-        location.public_send(part)
-      end
+      has_part = node.respond_to?(:location) && node.location.respond_to?(part)
+      node.location.public_send(part) if has_part
```

Comment:
Another approach to this could be:

```ruby
def self.source_range(node, part)
  [:location, part].reduce(node) do |object, method|
    object.respond_to?(method) ? object.public_send(method) : break
  end
end
```

I'm not saying this is better, just supplying an alternate approach for consideration.